delivery-service/src/main.rs

```diff
@@ -0,0 +1,308 @@
+        for (hash, key_package) in self.0.iter() {
+            encode_vec(VecSize::VecU16, buffer, &hash)?;
+            key_package.encode(buffer)?;
+        }
+        Ok(())
+    }
+
+    fn decode(cursor: &mut Cursor) -> Result<Self, CodecError> {
+        let length = u32::decode(cursor)?;
+        let mut key_packages = Vec::with_capacity(length as usize);
+        for _ in 0..length {
+            let hash = decode_vec(VecSize::VecU16, cursor)?;
+            let key_package = KeyPackage::decode(cursor)?;
+            key_packages.push((hash, key_package));
+        }
+        Ok(ClientKeyPackages(key_packages))
+    }
+}
+
+#[derive(Debug, Default, Clone)]
+pub(crate) struct ClientInfo {
+    client_name: String,
+    key_packages: ClientKeyPackages,
+    msgs: Vec<MLSMessage>,
+    welcome_queue: Vec<Welcome>,
+}
+
+impl Codec for ClientInfo {
+    fn encode(&self, buffer: &mut Vec<u8>) -> Result<(), CodecError> {
+        let client_name_bytes = self.client_name.as_bytes();
+        (client_name_bytes.len() as u16).encode(buffer)?;
+        buffer.extend_from_slice(&client_name_bytes);
+
+        (self.key_packages.0.len() as u16).encode(buffer)?;
+        for key_package in self.key_packages.0.iter() {
+            encode_vec(VecSize::VecU16, buffer, &key_package.0)?;
+            key_package.1.encode(buffer)?;
+        }
+        Ok(())
+    }
+
+    fn decode(cursor: &mut Cursor) -> Result<Self, CodecError> {
+        let client_name_length = u16::decode(cursor)?;
+        let client_name = std::str::from_utf8(cursor.consume(client_name_length.into())?)
+            .unwrap()
+            .to_string();
+
+        let mut key_packages = Vec::new();
+        let num_key_packages = u16::decode(cursor)?;
+        for _ in 0..num_key_packages {
+            let hash = decode_vec(VecSize::VecU16, cursor)?;
+            let key_package = KeyPackage::decode(cursor)?;
+            key_packages.push((hash, key_package));
+        }
+        Ok(Self::new(client_name, key_packages))
+    }
+}
+
+impl ClientInfo {
+    #[cfg(test)]
+    pub(crate) fn new(
+        client_name: String,
+        key_packages: Vec<(KeyPackageHash, KeyPackage)>,
+    ) -> Self {
+        Self {
+            client_name,
+            key_packages: ClientKeyPackages(key_packages),
+            msgs: Vec::new(),
+            welcome_queue: Vec::new(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+enum Message {
+    MLSMessage(MLSMessage),
+    Welcome(Welcome),
+}
+
+#[derive(Debug, Clone, PartialEq)]
+enum MLSMessage {
+    MLSCiphertext(MLSCiphertext),
+    MLSPlaintext(MLSPlaintext),
+}
+
+/// The return value for /msg/recv
+#[derive(Debug, PartialEq)]
+pub struct MessageList(Vec<Message>);
+
+impl Codec for Message {
+    fn encode(&self, buffer: &mut Vec<u8>) -> Result<(), CodecError> {
+        match self {
+            Message::MLSMessage(m) => match m {
+                MLSMessage::MLSCiphertext(m) => {
+                    0u8.encode(buffer)?;
```

Comment:
Create an enum for message type?

Reply:
yeah, would be a little cleaner. Will do.